File: packages/ui-swiftui/Sources/KinetixUI/Popover.swift
//
// Popover.swift — KinetixPopover.
//
// Mirrors packages/ui/src/components/popover.tsx. Wraps SwiftUI's native
// `.popover(isPresented:)` — anchor-relative positioning and outside-tap
// dismissal for free, the "reuse the platform machinery" call the
// Compose port made with Material3's `DropdownMenu`. `w-72` (288) is off
// the shared spacing scale.
//
// On iOS in a compact width the system adapts `.popover` to a sheet
// (arbitrary-content popovers as true popovers need
// `.presentationCompactAdaptation`, iOS 16.4+ — above this package's iOS
// 16.0 floor) — a documented platform behaviour, not tuned here.
//

import SwiftUI

public struct KinetixPopover<Anchor: View, Content: View>: View {
    @Environment(\.kinetixColors) private var colors

    @Binding private var isPresented: Bool
    private let anchor: Anchor
    private let popoverContent: Content

    public init(
        isPresented: Binding<Bool>,
        @ViewBuilder anchor: () -> Anchor,
        @ViewBuilder content: () -> Content
    ) {
        self._isPresented = isPresented
        self.anchor = anchor()
        self.popoverContent = content()
    }

    public var body: some View {
        anchor
            .popover(isPresented: $isPresented) {
                popoverContent
                    .padding(16) // spacing/4
                    .frame(minWidth: 288, alignment: .leading) // w-72
                    .background(colors.popover)
            }
    }
}

File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Popover.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.unit.dp

/**
 * KinetixPopover — mirrors `packages/ui/src/components/popover.tsx`. Built
 * on Material3's `DropdownMenu` (itself built on `Popup`) rather than a
 * hand-rolled anchored overlay — it already does anchor-relative
 * positioning and outside-tap dismissal, the same "reuse the platform
 * machinery" call [KinetixDialog]/[KinetixSlider] made. `anchor` renders
 * normally in the layout (taking the place of Radix's `PopoverTrigger`);
 * `content` renders inside the menu, restyled away from Material3's own
 * menu look to the Popover spec. `w-72` (288dp) isn't on the shared
 * `spacing_*` scale — hardcoded, same reasoning as `KinetixFab`'s
 * off-scale sizes. Side/align/offset auto-flip positioning is entirely
 * `DropdownMenu`'s own — not reproduced or tuned here.
 */
@Composable
fun KinetixPopover(
    visible: Boolean,
    onDismissRequest: () -> Unit,
    modifier: Modifier = Modifier,
    anchor: @Composable () -> Unit,
    content: @Composable () -> Unit,
) {
    val colors = KinetixColorScheme.current
    val shape = RoundedCornerShape(dimensionResource(R.dimen.radius_md))

    Box {
        anchor()
        DropdownMenu(
            expanded = visible,
            onDismissRequest = onDismissRequest,
            modifier = modifier
                .width(288.dp) // w-72, not on the shared scale
                .clip(shape)
                .background(colors.popover, shape)
                .padding(dimensionResource(R.dimen.spacing_4)),
        ) {
            CompositionLocalProvider(LocalContentColor provides colors.popoverForeground) {
                content()
            }
        }
    }
}
